apps/pkeyutl.c

```diff
@@ -313,14 +304,45 @@ int pkeyutl_main(int argc, char **argv)
         goto opthelp;
     }
 
+    pkey = get_pkey(kdfalg, inkey, keyform, key_type, passinarg, pkey_op, e);
+    if (pkey_op == EVP_PKEY_OP_SIGN || pkey_op == EVP_PKEY_OP_VERIFY) {
+        if (only_rawin(pkey)) {
+            if ((EVP_PKEY_is_a(pkey, "ED25519") || EVP_PKEY_is_a(pkey, "ED448"))
+                && digestname != NULL) {
+                BIO_printf(bio_err,
+                           "%s: -digest (prehash) is not supported with EdDSA\n", prog);
+                EVP_PKEY_free(pkey);
+                goto end;
+            }
+            rawin = 1; /* implied for Ed25519(ph) and Ed448(ph) and maybe others in the future */
+        }
+    } else if (rawin) {
+        BIO_printf(bio_err,
+                   "%s: -rawin can only be used with -sign or -verify\n", prog);
+        EVP_PKEY_free(pkey);
+        goto opthelp;
+    }
+    if (digestname != NULL && !rawin) {
+        BIO_printf(bio_err,
+                   "%s: -digest can only be used with -rawin\n", prog);
+        EVP_PKEY_free(pkey);
+        goto opthelp;
+    }
+
+    if (rawin && rev) {
+        BIO_printf(bio_err, "%s: -rev cannot be used with raw input\n", prog);
+        EVP_PKEY_free(pkey);
+        goto opthelp;
+    }
+
     if (rawin) {
         if ((mctx = EVP_MD_CTX_new()) == NULL) {
             BIO_printf(bio_err, "Error: out of memory\n");
+            EVP_PKEY_free(pkey);
             goto end;
         }
     }
-    ctx = init_ctx(kdfalg, &keysize, inkey, keyform, key_type,
-                   passinarg, pkey_op, e, engine_impl, rawin, &pkey,
+    ctx = init_ctx(kdfalg, &keysize, pkey_op, e, engine_impl, rawin, pkey,
```

Comment:
This is supposed to **take ownership** of `pkey`, but it can return `NULL` early on error, without freeing `pkey`, leaving it dangling.

Reply:
Oh, good spotting!
I've now fixed all those potential memore leak cases in `init_ctx()`.